### PR TITLE
adding a beta parameter to the smooth_l1 loss fn

### PR DESCRIPTION
**Pull request by bdhirsh**
PR for issue: https://github.com/pytorch/pytorch/issues/16357

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44433 adding a beta parameter to the smooth_l1 loss fn**

Not entirely sure why, but changing the type of beta from `float` to `double in autocast_mode.cpp and FunctionsManual.h fixes my compiler errors, failing instead at link time

fixing some type errors, updated fn signature in a few more files

removing my usage of Scalar, making beta a double everywhere instead

Differential Revision: [D23636720](https://our.internmc.facebook.com/intern/diff/D23636720)

### CODE REVIEW

File: aten/src/ATen/native/cpu/PointwiseOpsKernel.cpp
Comment (bdhirsh) [inline aten/src/ATen/native/cpu/PointwiseOpsKernel.cpp:65]: the smooth l1 loss when |X| < 1 now divides by the beta parameter, so the derivative does as well.

**Comment by bdhirsh** [comment]
Making beta a double rather than a Scalar, since it's used in comparison- something that doesn't really make sense for complex types. See comment here (sorry for the duplicate PR) https://github.com/pytorch/pytorch/pull/44197#pullrequestreview-483654166

File: aten/src/ATen/native/PointwiseOps.h
Comment (bdhirsh) [inline aten/src/ATen/native/PointwiseOps.h:14]: lmk if you think the name should be more general, or if I just shouldn't use an alias here. Right now smooth_l1_backwards looks like the only point wise op here that takes in two scalar params

File: torch/csrc/api/include/torch/nn/functional/loss.h
Comment (bdhirsh) [inline torch/csrc/api/include/torch/nn/functional/loss.h:321]: it's not 100% clear to me if I need to change this, since this seems independent of the kernel code

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (bdhirsh) [inline torch/csrc/autograd/FunctionsManual.cpp:1089]: not 100% on how beta should be used in the double_backwards of smooth_l1_loss

**Review by mrshenli — COMMENTED** [review]
Thanks for adding this!

We might also need to change the [SmoothL1Loss](https://github.com/pytorch/pytorch/blob/fab012aa28e58fdbf49e7da29cd7f5bff90f6bdb/torch/nn/modules/loss.py#L758) class to use this new function API, and update its doc string as well. This can come in a followup PR.

File: aten/src/ATen/native/Loss.cpp
Comment (gchanan) [inline aten/src/ATen/native/Loss.cpp:307]: this looks wrong, although it was wrong before your patch.  You can't just assign the result to the C++ reference (the reference type is bogus); the semantics are to rewrite the _memory_ in `result`.

File: aten/src/ATen/native/Loss.cpp
Comment (bdhirsh) [inline aten/src/ATen/native/Loss.cpp:308]: @gchanan I updated smooth_l1_loss_out to something similar to what MSE does:
- make a TensorIterator that puts the smooth l1 loss output in a temporary tensor
- use the existing `<reduction>_out` functions to put the output of the reduction in the out var.

My educated guess as to why we need the temporary variable is because `result` has a different tensor shape than the tensor output of the actually kernel function when the caller is using reductions.

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp
Comment (gchanan) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:515]: this looks like the math is going to be different for the vectorized vs not vectorized case.  In particular, won't the math be done in double here (if scalar_t is float), whereas it will be float in the vectorized case?

File: aten/src/ATen/native/cuda/BinaryMiscOpsKernels.cu
Comment (gchanan) [inline aten/src/ATen/native/cuda/BinaryMiscOpsKernels.cu:26]: similar here, isn't the math in double now?

File: aten/src/ATen/native/cuda/PointwiseOpsKernel.cu
Comment (gchanan) [inline aten/src/ATen/native/cuda/PointwiseOpsKernel.cu:33]: this looks like a typo.  Also the same double issue.

File: torch/nn/functional.py
Comment (gchanan) [inline torch/nn/functional.py:2574]: it doesn't look like this is actually used?

File: torch/nn/functional.py
Comment (gchanan) [inline torch/nn/functional.py:2581]: this type doesn't look optional.

File: torch/nn/functional.py
Comment (bdhirsh) [reply torch/nn/functional.py:2574]: yes- wasn't sure originally if I could just delete it. I'll take it out now.

File: torch/nn/functional.py
Comment (bdhirsh) [reply torch/nn/functional.py:2581]: Fixing, thanks.

File: aten/src/ATen/native/cuda/PointwiseOpsKernel.cu
Comment (bdhirsh) [reply aten/src/ATen/native/cuda/PointwiseOpsKernel.cu:33]: I finally got cuda to build successfully this morning and point out this error to me :)

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp
Comment (bdhirsh) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:515]: yes, thanks for pointing that out. I'm converting beta from double to Vec256<scalar_t> near the top.

File: aten/src/ATen/native/cuda/BinaryMiscOpsKernels.cu
Comment (bdhirsh) [reply aten/src/ATen/native/cuda/BinaryMiscOpsKernels.cu:26]: In general, do we want scalar inputs to our kernels to have the same precision as the dtype of the tensor(s) they operate on? In that case it sounds a little harder since we don't want beta to take on a complex type:
- When scalar_t is a complex, use beta with double precision
- Otherwise, use beta as a scalar_t

Or would another option be just to directly make beta a float rather than a double (which I think would require reducing precision when we go from python to C++).

**Comment by gchanan** [comment]
For posterity: https://github.com/pytorch/pytorch/issues/16357 is the issue (you should include it in the PR description).

**Comment by bdhirsh** [comment]
@JackCaoG Hi Jack, I think that this PR requires a change on the XLA side- I'm updating the existing `smooth_l1_loss` operator to take in an optional `beta` param, and I noticed a failing XLA build from the PR: https://app.circleci.com/pipelines/github/pytorch/pytorch/213829/workflows/0fb2f612-971c-4fbf-aca5-974a1d85a11b/jobs/7459944.

`AtenXlaType function missed override: Tensor smooth_l1_loss(const Tensor& self, const Tensor& target, int64_t reduction);`

Let me know if I can help!

**Comment by bdhirsh** [comment]
@robieta Posting the perf results from the torch timer after your walkthrough of how to use it. I used the torch timer to measure calls to `smooth_l1_loss` before/after the change, with 4 random tensors for 10 seconds each. and I don't think any regression here is noticeable (I noticed that the IQR's are pretty low for a given run, but the medians tended to move around a little more when I tried running each test multiple times). We'd expect maybe a small increase since the `smooth_l1_loss` call now requires an extra multiply/divide, but nothing here looks like a big change. Let me know if anything looks suspicious/wrong to you!

(I think that the warnings in the screenshot below are fine, since we only really care about the performance and not correctness of the op call).

Before:
(cpu)
image

(cuda)
image


After:
(cpu)
image

(cuda)
image

**Comment by ezyang** [comment]
Yes perf looks within noise. LGTM!

**Comment by JackCaoG** [comment]
> @JackCaoG Hi Jack, I think that this PR requires a change on the XLA side- I'm updating the existing `smooth_l1_loss` operator to take in an optional `beta` param, and I noticed a failing XLA build from the PR: https://app.circleci.com/pipelines/github/pytorch/pytorch/213829/workflows/0fb2f612-971c-4fbf-aca5-974a1d85a11b/jobs/7459944.
> 
> `AtenXlaType function missed override: Tensor smooth_l1_loss(const Tensor& self, const Tensor& target, int64_t reduction);`
> 
> Let me know if I can help!

Thanks for the heads-up, I will make the corresponding change on the XLA side.

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (mrshenli) [inline torch/csrc/autograd/FunctionsManual.cpp:919]: can beta be 0 here?

File: aten/src/ATen/native/cpu/PointwiseOpsKernel.cpp
Comment (mrshenli) [inline aten/src/ATen/native/cpu/PointwiseOpsKernel.cpp:66]: looks like if `x == beta == 0`, we could reach here, and then trigger division by 0?

File: torch/testing/_internal/common_nn.py
Comment (mrshenli) [inline torch/testing/_internal/common_nn.py:3622]: Does this also need to handle divide by 0 case?

File: aten/src/ATen/native/cpu/PointwiseOpsKernel.cpp
Comment (mrshenli) [reply aten/src/ATen/native/cpu/PointwiseOpsKernel.cpp:66]: I might miss sth, but looks like this is different from the match given in the doc. 

```
0.5 (x_i - y_i)^2 / beta, & \text{if } |x_i - y_i| < beta 
```

Should the three condition be `x <= -beta`, `x >= beta`, and otherwise?

**Review by mrshenli — COMMENTED** [review]
If the concern of division by 0 is a real, let's add a test to check where `|x - y| == 0`

File: torch/nn/modules/loss.py
Comment (mrshenli) [inline torch/nn/modules/loss.py:813]: this makes beta a class attribute instead of an instance attribute?

File: aten/src/ATen/native/cpu/PointwiseOpsKernel.cpp
Comment (bdhirsh) [reply aten/src/ATen/native/cpu/PointwiseOpsKernel.cpp:66]: yep you're right, thanks. And using `x <= -beta` and `x >= beta` should fix the division by 0 issue.

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (bdhirsh) [reply torch/csrc/autograd/FunctionsManual.cpp:919]: yep, good catch. I'll try rewriting it in a way to prevent divide by zeros.

File: torch/testing/_internal/common_nn.py
Comment (mrshenli) [inline torch/testing/_internal/common_nn.py:889]: this needs to be added to `new_module_tests`?

File: torch/testing/_internal/common_nn.py
Comment (mrshenli) [inline torch/testing/_internal/common_nn.py:897]: this won't trigger zero loss, right? If we want to test zero loss, will using `t.clone()` as input work?

File: aten/src/ATen/native/cpu/PointwiseOpsKernel.cpp
Comment (mrshenli) [reply aten/src/ATen/native/cpu/PointwiseOpsKernel.cpp:67]: I might be wrong. If it divides by beta in the forward pass, should we multiply the grad by beta in the backward pass?

File: torch/testing/_internal/common_nn.py
Comment (mrshenli) [inline torch/testing/_internal/common_nn.py:3639]: since beta is 0, we can remove the 0.5 * beta from the formula?

File: aten/src/ATen/native/cpu/PointwiseOpsKernel.cpp
Comment (mrshenli) [reply aten/src/ATen/native/cpu/PointwiseOpsKernel.cpp:67]: Chatted with @bdhirsh offline. I was wrong. Brian's current implementation is correct.

File: torch/testing/_internal/common_nn.py
Comment (bdhirsh) [reply torch/testing/_internal/common_nn.py:897]: Two things:

1) I started trying to modify the backwards kernels to avoid/check for divide-by-zeros, but I realized that setting beta=0 is really just the same thing as using l1_loss. I updated the definitions in `Loss.cpp` to check for that case and just call out to their corresponding `l1_loss` functions directly.

2) I added a test to the suite that uses beta=0, but I'm not testing for the case that the two tensors are the same. That's because it doesn't look like that test will play well with this test suite. The test suite asserts the logic by computing the analytical jacobian of the loss (using my backwards) as well as the numerical jacobian (using (f(x+e)-f(x-e)/(2*e))), but that logic breaks down when one of the functions isn't continuous. In this case, the second derivative of smooth_l1_loss when beta=0 isn't continuous- it's equal to infinity at x=0, and 0 everywhere else.

**Comment by bdhirsh** [comment]
Hey @JackCaoG, I see that the PR on the XLA side has approval, so I'm going to start trying to merge this soon. Thanks for helping!

**Comment by JackCaoG** [comment]
> Hey @JackCaoG, I see that the PR on the XLA side has approval, so I'm going to start trying to merge this soon. Thanks for helping!

sounds good!

**Review by mrshenli — APPROVED** [review]
LGTM! Let's land after the XLA PR, so that we won't break master.

**Comment by JackCaoG** [comment]
> LGTM! Let's land after the XLA PR, so that we won't break master.

You guys can go first, in xla side I can simply merge the pr after this one merged(I think pytorch merge takes some time).

File: aten/src/ATen/native/native_functions.yaml
Comment (gchanan) [inline aten/src/ATen/native/native_functions.yaml:6558]: nit: you shouldn't need to provide a default parameter -- we don't expose these backwards functions to python, they should only be called via autograd.  And you'll notice that reduction doesn't have a default parameter.

The parts I'm not sure about here are if TorchScript would let you serialize this function (it shouldn't) and/or if the backwards compatibility test would complain.

File: torch/nn/modules/loss.py
Comment (gchanan) [inline torch/nn/modules/loss.py:774]: the above isn't accurate anymore ""Creates a criterion that uses a squared term if the absolute
    element-wise error falls below 1 and an L1 term otherwise.	    element-wise error falls below 1 and an L1 term otherwise."

Comment (gchanan) [reply torch/nn/modules/loss.py:774]: we should also document the beta < 0.

**Pull request closed without merging**
